crates/oxc_linter/src/rules/react/style_prop_object.rs

```diff
@@ -0,0 +1,428 @@
+use oxc_ast::{
+    ast::{
+        Argument, Expression, JSXAttribute, JSXAttributeItem, JSXAttributeName, JSXAttributeValue,
+        JSXElementName, ObjectPropertyKind,
+    },
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{
+    ast_util::get_declaration_of_variable,
+    context::{ContextHost, LintContext},
+    rule::Rule,
+    utils::is_create_element_call,
+    AstNode,
+};
+
+fn style_prop_object_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Style prop value must be an object")
+        .with_help("Make sure the 'style' prop value is an object")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct StylePropObject(Box<StylePropObjectConfig>);
+
+#[derive(Debug, Default, Clone)]
+pub struct StylePropObjectConfig {
+    allow: Vec<String>,
+}
+
+impl std::ops::Deref for StylePropObject {
+    type Target = StylePropObjectConfig;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    /// Require that the value of the prop style be an object or a variable that is an object.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```jsx
+    /// <div style="color: 'red'" />
+    /// <div style={true} />
+    /// <Hello style={true} />
+    /// const styles = true;
+    /// <div style={styles} />
+    ///
+    /// React.createElement("div", { style: "color: 'red'" });
+    /// React.createElement("div", { style: true });
+    /// React.createElement("Hello", { style: true });
+    /// const styles = true;
+    /// React.createElement("div", { style: styles });
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```jsx
+    /// <div style={{ color: "red" }} />
+    /// <Hello style={{ color: "red" }} />
+    /// const styles = { color: "red" };
+    /// <div style={styles} />
+    ///
+    /// React.createElement("div", { style: { color: 'red' }});
+    /// React.createElement("Hello", { style: { color: 'red' }});
+    /// const styles = { height: '100px' };
+    /// React.createElement("div", { style: styles });
+    /// ```
+    StylePropObject,
+    suspicious
+);
+
+fn is_invalid_expression(expression: Option<&Expression>, ctx: &LintContext<'_>) -> bool {
+    let Some(expression) = expression else {
+        return false;
+    };
+
+    match expression {
+        Expression::StringLiteral(_) | Expression::BooleanLiteral(_) => true,
+        Expression::Identifier(ident) => {
+            let Some(node) = get_declaration_of_variable(ident, ctx) else {
+                return false;
+            };
+
+            let AstKind::VariableDeclarator(var_decl) = node.kind() else {
```

Comment:
Consider checking `FormalParameter`s as well

Reply:
I don't have a great grasp of the AST yet, but can FormalParameters also be something other than function parameters?

Reply:
They're used in type parameters, but I don't think that's relevant here.